src/set_relation/set_relation.cc

```diff
@@ -3574,7 +3581,153 @@ Relation* Relation::boundDomainRange()
+#pragma mark -
+/*************** VisitorSuperAffineSet *****************************/
+// Vistor Class used in SuperAffineSet
+// Used in traversing a Set/Relation to replace UFCalls with symbolic constants
+// We will build up a affineSet (or affineRelation), term by term.
+// And if a term is UFCall, we will convert it to symbolic constant,
+// which is pre-computed and stored in the ufcmap
+class VisitorSuperAffineSet : public Visitor {
+  private:
+         UFCallMap* ufcmap;
+         Set* affineSet;
+         Relation* affineRelation;
+         Conjunction* affineConj;
+         bool rel;         // To know whether we are traversing a Relation
+  public:
+         VisitorSuperAffineSet(UFCallMap* imap){ ufcmap = imap;}
+         //! Builds up an affineExp and adds it to affineConj
+         void preVisitExp(iegenlib::Exp * e);
+         //! Initializes an affineConj
+         void preVisitConjunction(iegenlib::Conjunction * c);
+         //! adds the current affineConj to affineSet (affineRelation)
+         void postVisitConjunction(iegenlib::Conjunction * c);
+         //! Initializes affineSet and rel
+         void preVisitSet(iegenlib::Set * s);
+         //! Initializes affineRelation and rel
+         void preVisitRelation(iegenlib::Relation * r);
+
+         Set * getSet(){ return affineSet; }
+         Relation * getRelation(){ return affineRelation; }
+};
+
+void VisitorSuperAffineSet::preVisitExp(iegenlib::Exp * e)
+{
+    //! If this is a argument to a UFCall, we don't want to modify it. 
+    if( e->isExpression() ){
+        return;
+    }
+
+    Exp* affineExp = new Exp();
+
+    //! We iterate over terms in Exp, if the term is not a UFCall
+    //  then we just add it to our affine set.
+    //  On the other hand, We need to turn UFCalls into
+    //  symbolic constants to make an affine set.
+    std::list<Term*> terms = e->getTermList();
+    for(std::list<Term*>::iterator i=terms.begin(); i != terms.end(); i++){
+       if( !( (*i)->isUFCall() ) ){
+           affineExp->addTerm( (*i)->clone() );
+       }
+       else{
+           UFCallTerm* cufc = dynamic_cast<UFCallTerm*>(*i);
+           cufc = dynamic_cast<UFCallTerm*>(cufc->clone());
+           int cof = cufc->coefficient();
+           cufc->setCoefficient(1);
+           std::string SymConst = ufcmap->find( cufc ); 
+           VarTerm* vt = new VarTerm( cof , SymConst );
+           affineExp->addTerm( vt );
+           delete cufc;
+       }
+    }
+
+    if( e->isInequality() ){
+        affineExp->setInequality();
+        affineConj->addInequality( affineExp );
+    }
+    else if( e->isEquality() ){
+        affineExp->setEquality();
+        affineConj->addEquality( affineExp );
+    }
+}
+
+void VisitorSuperAffineSet::preVisitConjunction(iegenlib::Conjunction * c)
+{
+    affineConj = new Conjunction( c->getTupleDecl() );
+    affineConj->setinarity( c->inarity() );
+}
+
+void VisitorSuperAffineSet::postVisitConjunction(iegenlib::Conjunction * c)
```

Comment:
Save the new conjunctions in a list in the visitor and then add in the Set or Relation as appropriate.